### PR TITLE
Simplify caching and storage for queries

### PR DESCRIPTION
**Pull request by jyn514**
I highly recommend reviewing commit-by-commit; each individual commit is quite small but it can be hard to see looking at the overall diff that the behavior is the same. Each commit depends on the previous.

r? @cjgillot

### CODE REVIEW

**Comment by jyn514** [comment]
The single large regression is in llvm, and there are many smaller improvements. I think this is just a change in codegen units that will go away on any further small change.

File: compiler/rustc_macros/src/query.rs [hunk 1]
```diff
@@ -286,12 +254,10 @@ fn add_query_description_impl(query: &Query, impls: &mut proc_macro2::TokenStrea
                 #expr
             }
 
-            #try_load_from_disk
+            const TRY_LOAD_FROM_DISK: Option<fn(QueryCtxt<'tcx>, SerializedDepNodeIndex) -> Option<Self::Value>>
+                = Some(crate::plumbing::try_load_from_disk::<Self::Value>);
```
Comment (cjgillot) [inline compiler/rustc_macros/src/query.rs:258]: I'm not really fond of referencing code from `rustc_query_impl` here.
Could we make `TRY_LOAD_FROM_DISK` a `bool`, and have `rustc_query_impl` pass the function pointer, or does it cause an issue with the where clause on `V: Decodable`?
Other possibility: add a `cache` modifier to be read by a macro in `rustc_query_impl` and avoid instantiating an uncallable function?

Comment (cjgillot) [reply compiler/rustc_macros/src/query.rs:258]: r=me if impossible / too impractical.

Comment (jyn514) [reply compiler/rustc_macros/src/query.rs:258]: > Could we make `TRY_LOAD_FROM_DISK` a `bool`, and have `rustc_query_impl` pass the function pointer, or does it cause an issue with the where clause on `V: Decodable`?

The where clause is exactly why I ended up with the current function, yeah. I didn't try to hard to fix it - I'll take a look this evening and see what I can do.

Comment (cjgillot) [reply compiler/rustc_macros/src/query.rs:258]: Other possibility: get rid of the `TRY_LOAD_FROM_DISK` constant, and use the same device as we use with `no_hash/hash_results`: a word in the modifiers, and a macro to select on it.

Comment (jyn514) [reply compiler/rustc_macros/src/query.rs]: > Other possibility: get rid of the TRY_LOAD_FROM_DISK constant, and use the same device as we use with no_hash/hash_results: a word in the modifiers, and a macro to select on it.

this is an excellent idea, done :)

File: compiler/rustc_query_impl/src/plumbing.rs [hunk 2]
```diff
@@ -420,8 +447,7 @@ macro_rules! define_queries {
                     hash_result: hash_result!([$($modifiers)*]),
                     handle_cycle_error: |tcx, mut error| handle_cycle_error!([$($modifiers)*][tcx, error]),
                     compute,
-                    cache_on_disk,
-                    try_load_from_disk: Self::TRY_LOAD_FROM_DISK,
+                    try_load_from_disk: if cache_on_disk { Self::TRY_LOAD_FROM_DISK } else { None },
```
Comment (cjgillot) [inline compiler/rustc_query_impl/src/plumbing.rs:450]: I think we can even put `should_ever_cache_on_disk!([$($modifiers)*])` here, and remove the `TRY_LOAD_FROM_DISK` associated constant.

Comment (jyn514) [reply compiler/rustc_query_impl/src/plumbing.rs]: great idea, done :)

**Comment by nnethercote** [comment]
The `deep-vector debug full` regression could be noise. Even if it's not, a single run of a secondary benchmark doesn't matter much, esp. given there are many small improvements across many other benchmarks.

@rustbot label: +perf-regression-triaged

**Pull request merged**